refcounted_he_(new|fetch)_pvn: Don't roll-own code

khwilliamson opened the pull request:
<!--
A good description should explain the problem the pull request addresses
and give context to the reviewers to aid them in their reviews.
-->
The function bytes_from_utf8() already does what what these two instances of duplicated code do.

<!--
Significant changes to Perl must be documented in perldelta.

Consider if the changes in this pull request are worthy of a perldelta
entry, then pick the appropriate line below and remove the others.
-->
---------------------------------------------------------------------------------

* This set of changes does not require a perldelta entry.

tonycoz commented:
My only issue with this is bytes_from_utf8() always pays the cost of an allocation, even when all of the characters are invariants, while the original code only pays that cost if an invariant is found.

khwilliamson commented:
I don't thing these hv functions have any business re-implementing utf8 translations.  It's unfortunate the API is the way it is.  I first tried the equivalent function `utf8_to_bytes` which is destructive of the input.  We know that the translated version will be no longer, and likely shorter than the original.  So that should work.  But I kept getting segfaults with what appeared to be valid writes.  I came to the conclusion it must be writing to some shared or read-only memory.

So what to do.  What strikes me immediately is to change the len parameter from STRLEN to SSize_t.  I think we can get away with that without affecting existing code; but if I'm wrong this won't fly.  If that is feasible, then a negative length would signal the function to just return the input if there are no variants.  The caller would check if the return pointer is the same as the passed one.

khwilliamson commented:
Another option is to convert `bool *is_utf8p` to an enum with an extra value that says turn off this flag and return the original source if everything is an invariant

khwilliamson commented:
That's not going to work either.  How about a new function `Size_t utf8_to_bytes_flags(U8 ** s, int type)`

- Returns 0 if  **s could not be converted; otherwise returns the number of bytes in the converted version.  Does nothing if **s doesn't need conversion; otherwise *s will point to the converted version.  By default, the conversion is destructive, meaning *s is converted itself.  The type is an enum with three states:
-  a) destructive 
-  b) non-destructive, so that new memory is allocated for the conversion should there need to be converting.
- c) non-destructive plus mortalized, so that any new memory will be deallocated without the caller needing to deal with it.

khwilliamson commented:
I have worked on this, and come with the following API proposal.   Feedback welcome

    "utf8_to_bytes_type"
    "utf8_to_bytes"
    "bytes_from_utf8"
        NOTE: "utf8_to_bytes" is experimental and may change or be removed
        without notice.

        These each convert a string encoded as UTF-8 into the equivalent
        native byte representation, if possible.

        "utf8_to_bytes_type" has the more modern API, and is the easiest to
        use. On input, "s_ptr" is a pointer to the string to be converted
        (so that the first byte will be at *sptr[0]), and *lenp is its
        length.

        It returns non-zero if the end result is native bytes; zero if there
        are code points in the string not expressible in native byte
        encoding. In many situations, treating the return as a boolean is
        sufficient, but it actually returns an enum "xxx" so you can tease
        apart the reason it succeeded. "noop" means that the input already
        was in bytes, so no action was necessary. "converted" means that the
        conversion was successful. "cant_convert" is a synonym you can use
        for 0 or "false".

        In all cases, *s_ptr and *lenp will have correct and consistent
        values, unchanged if the return is either 0 or "noop"; otherwise
        updated as necessary.

        If the return is "converted" and "type" is 0 (or the equivalent
        "overwrite"), the converted value overwrote the input string. *s_ptr
        will be unchanged, but *lenp will be updated to its new (shortened)
        length.

        If the return is "converted" and "type" is "preserve", the original
        string is never changed. Instead a new "NUL"-terminated string is
        allocated, and *s_ptr is changed to point to that new memory. The
        caller is responsible for freeing that memory. "type" can also be
        set to "use_temp". In this case the original string is also
        preserved, and the converted string placed in newly allocated
        memory, but that will have been marked for automatic destruction (by
        using "SAVEFREEPV").

        Note that the caller has to free the memory at *s_ptr if and only if
        it called this function with "type" set to "preserve", and the
        function returned "converted".

        Upon successful return, the number of variants in the string can be
        computed by having saved the value of *lenp before the call, and
        subtracting the after-call value of *lenp from it. This is also true
        for the other two functions described below.

        Plain "utf8_to_bytes" also converts a UTF-8 encoded string to bytes,
        but there are more glitches that the caller has to be prepared to
        handle.

        The input string is passed with one less indirection level, "s".

        If the conversion was successful or a noop
            The function returns "s" (unchanged) and *lenp will contain the
            correct length.

        If the conversion failed
            The function returns NULL and sets *lenp to -1, cast to
            "STRLEN". This means that you will have to use a temporary
            containing the string length to pass to the function if you will
            need the value afterwards.

        "bytes_from_utf8" also converts a potentially UTF-8 encoded string
        "s" to bytes. It preserves "s", allocating new memory for the
        converted string.

        In contrast to the other two functions, the input string to this one
        need not be UTF-8. If not, the caller has set *is_utf8p to be
        "false", and the function does nothing, returning the original "s".

        Also do nothing if there are code points in the string not
        expressible in native byte encoding, returning the original "s".

        Otherwise, *is_utf8p is set to 0, and the return value is a pointer
        to a newly created string containing the native byte equivalent of
        "s", and whose length is returned in *lenp, updated. The new string
        is "NUL"-terminated. The caller is responsible for arranging for the
        memory used by this string to get freed.

        The major problem with this function is that memory is allocated
        and filled even when no conversion was necessary..

            U8         utf8_to_bytes_type(      U8 **s_ptr, STRLEN *lenp,
                                                U32 type)
            U8 *       utf8_to_bytes     (      U8 *s, STRLEN *lenp)
            U8 *  Perl_utf8_to_bytes     (pTHX_ U8 *s, STRLEN *lenp)
            U8 *       bytes_from_utf8   (      const U8 *s, STRLEN *lenp,
                                                bool *is_utf8p)
            U8 *  Perl_bytes_from_utf8   (pTHX_ const U8 *s, STRLEN *lenp,
                                                bool *is_utf8p)

khwilliamson commented:
> One question, is there any way to undo in perlapi SAVEFREEPV? like for sv_usepvn_flags?

I don't believe so

khwilliamson committed:
utf8.c: White-space only

Outdent and reflow some comments and code in preparation for them to be
moved out of the loop

khwilliamson committed:
utf8.c: Move declaration to first use

khwilliamson committed:
utf8_to_bytes() Move failure code out of loop

This is for clarity.  All this very-unlikely-to-be-used code was in the
middle of what is really going on, creating a distraction.

khwilliamson committed:
utf8_to_bytes: Refactor loop

The previous version did not make sure that it wasn't reading beyond the
end of the buffer in all cases, and the first pass through the input
string already ruled out it having most problems.  Thus we don't need
the full generality here of the macro UTF8_IS_DOWNGRADEABLE_START; and
this simplifies things

khwilliamson committed:
utf8_to_bytes: Update and fix comments.

These were misleading.  On ASCII platforms, many calls to this function
won't use the per-word algorithm.  That's only done for long-enough
strings.

khwilliamson committed:
utf8_to_bytes: Rename variable

The new name, s0, is used in more other places for this meaning, and is
more descriptive.

khwilliamson committed:
Add preliminary utf8_to_bytes_()

This is an internal function, designed to be an extension of
utf8_to_bytes(), with a slightly different API.  This commit just adds
it and calls it from just utf8_to_bytes.

Future commits will extend this API.

khwilliamson committed:
utf8_to_bytes_: Add const

This variable should not be being changed by the function

khwilliamson committed:
Document new utf8_to_bytes() variants

khwilliamson committed:
utf8_to_bytes_: Add argument, macro

The argument is currently unused.  The macro is a public facing API that
calls this function with the correct argument

khwilliamson committed:
utf8_to_bytes_: Slight refactor

This makes the next commit smaller

khwilliamson committed:
utf8_to_bytes_: Calculate needed malloc size

Prior to this commit, the size malloced was just the same as the length
of the input string, which is a worst case scenario.  This commit
changes so the new pass through the input (introduced in the previous
commit) also calculates the needed length.

The additional cost of doing this is minimal.  It has advantages on a
very long string with lots of sequences that are convertible.

khwilliamson committed:
utf8_to_bytes_: Add non-destructive write option

This causes this function to be able to both overwrite the input, and to
instead create new memory.  It changes bytes_from_utf8() to use this new
capability instead of being a near duplication of the core code of this
function.

Prior to this commit, bytes_from_utf8() just allocated memory the size
of the original string, and started copying into it.  When it came to a
sequence that wasn't convertible, it stopped, and freed up the copy.
The new behavior has it checking first before the malloc that the string
is convertible.  That has the advantage that there is no malloc without
being sure it will be useful; but the disadvantage that there is an
extra pass through the input string, but that pass is per-word.

The next commit will introduce another advantage.

Thanks to Tony Cook for the 'free_me' idea

khwilliamson committed:
utf8_to_bytes_: Add ability to return a mortalized pv

This is a non-destructive conversion of the input into native bytes, and
with any new memory required set for destruction via SAVEFREEPV.  This
allows the caller to not have to be concerned at all if memory was
created or not.

A new macro is created that calls this internal function with the
correct parameter to force this behavior.

tonycoz commented:
Rather than requiring the caller to decode the return value to decide whether the string should be released, perhaps add another pointer to pointer:

```
U8         utf8_to_bytes_type(      U8 **s_ptr, STRLEN *lenp,
                                            U32 type, void **free_me)
```
utf8_to_bytes_type() would set `*free_me` to NULL and iff the caller needs to free ortake ownership of the PV, it sets `*free_me` to that pointer.  This way a caller can just:

```
U8 *orig = ...;
STRLEN len = ...;
void *free_me; // possibly U8 * instead
utf8_bytes_to_type(&orig, &len, preserve, &free_me);
// process orig for len bytes
Safefree(free_me); // Safefree() handles NULL like free()
```

To make the call even less error prone, supply the original string pointer/length by value to avoid the caller having to remember to initialize them:
```
U8 *outstr;
STRLEN outlen;
void *free_me;
utf8_bytes_to_type(&outstr, &outlen, origstr, origlen, preserve, &free_me);
// process outstr for outlen bytes
Safefree(free_me);
```

If utf8_to_bytes_type() returns or accepts an enum, its return and parameter types should reflect that.

Please document the enum values for the `type` parameter and return types as lists.

Hopefully the the enum names will decorated to avoid name conflicts.

> One question, is there any way to undo in perlapi SAVEFREEPV? like for sv_usepvn_flags?

It's not needed here, since you can call with type = preserve.

But in general there's no easy way to do that (like release() for [std::unique_ptr](https://en.cppreference.com/w/cpp/memory/unique_ptr)in C++.

It would be nice to have that for both this and sv_2mortal(), and there's at least one place that manually does this for mortal SVs.

khwilliamson commented:
I forgot about this pull request when I created a more refined version of the utf8_to_bytes proposal.  I did that in #22703, now closed.  And this has been updated to the latest.

One thing I realized when converting existing calls to the new way was that there was a problem with const.  The same function now has to be callable in cases where it both changes the input, and is required to not change the input, depending on a flag.  To solve this, I made the function internal, and created macros that don't cast away const except when the actual function isn't going to change it.  This protects the caller from calling the destructive version with a const string.

The macros are currently named utf8_to_bytes_overwrite() utf8_to_bytes_new_pv(), and utf8_to_bytes_temp_pv().  Better name ideas welcome

These hide from the user the new enum parameter that the internal function is called with.

Using @tonycoz idea of an extra parameter that the internal function sets to indicate that there is a need for freeing the converted string means the enum return type can be gotten rid of, and the functions return bool.  Only the new-pv case actually needs that, so having the macros as the public interface hides this parameter except when needed.

I had this parameter as a pointer to a bool.  @tonycoz idea of making it point to the memory to be freed is a better idea, but in all cases in the core, that doesn't simplify things;  they each need to do more things than free the memory when the function has created that memory.

Included in this pr are the conversions to use the new function in all the core files. These are intended to not be actually merged at the same time as the rest, but showed how this would work out in practice.

tonycoz commented on utf8.h:
```diff
@@ -1305,8 +1305,15 @@ typedef enum {
+ * wrongly gets used */
 #define Perl_utf8_to_bytes_overwrite(mTHX, s, l)                            \
-        Perl_utf8_to_bytes_(aTHX_ s, l, PL_utf8_to_bytes_overwrite)
+        Perl_utf8_to_bytes_(aTHX_ s, l, (bool *) 1,                         \
+                                  PL_utf8_to_bytes_overwrite)
```
This cast `(bool *) 1` is unsafe.

Such casts are implementation defined, and may result in an invalid pointer, use (which the function call itself does) of which may result in undefined behaviour.

khwilliamson replied:
I was too lazy to look up how to use INT2PTR, which would have been less effort than what it ended up costing me.  I changed to use that and also to using U8* instead of bool.  Presumably that is no longer UB

tonycoz replied:
No, it's still invalid.

khwilliamson replied:
Does that mean `INT2PTR` is unsafe.  I'm expecting bad things to happen should this ever get exercised.  It may be undefined behavior, but I can't imagine it would be worse than a segfault.

(I worked on a machine once that had an instruction that should have been labelled `BLOW FUSE` instead of what it was supposed to do)

tonycoz replied:
INT2PTR() or any cast of an integer to a pointer is unsafe unless that integer is zero, or the result of casting a pointer to some object (or a one past the end pointer) to an integer.

Now, an implementation can allow such casts, and they often do, but it is unsafe going by the standard.

Frankly, if you want to pass a pointer to not-an-object, pass NULL, that is what is it for.

bulk88 replied:
"Handles" and "opaque" pointers less than +/- 4096 away from signed integer 0, or +/- 128 or +/- 8 away from zero are not going to ever be de-referenceable values. Opaque pointers -1 and +1  are alignment faults anyways in a court room trial. Harvard architecture, 65KB pointers and segmented memory, or 16-bit Windows OS allowing the user to randomly remove a floppy that has current process/current executing .dll/.so/.exe!!! without a crash!!!!, only a temp popup box.

Its not coming back. And HPUX's NULL is a valid readable no-SEGV address in ISO C isn't coming back either.

Only danger is going from 0x40/-64 or +64 to 0x80  between byte 1 and byte 2, brings up danger of accidental sign extension of your flag, in a sloppy codebase. Throw ontop of that a "clever" `&` to test for pointers instead of <>, and you have a bug ticket.

Only real danger of 21st century is getting negative malloc pointers back from libc. Some Unix RISC CPU/OS companies decided user-mode is always negative malloc pointers https://github.com/Tencent/rapidjson/issues/1596 

And "2 free bytes" with AMD64 in 64-bit mode is just that (+Intel), AMD64 in 64 bit mode. Other older OSes spread out the default per proc categories of VM pages wide enough, you didn't get a solid USHORT16 of bits to use for personal intra-process uses (encrypting pointers, CRC32s, hash key hints, network forwarding path/leg hints).

edit; FWIW, I've never in my life used the free "2 bytes" on AMD64.  I will always convert the API to U32/U16/U8 offsets (abs addr diff) or indexs (unit of 1, need scaling). And in general on ancient windows/current POSIX,  you'll save private commit memory since your removing shared lib pointer "relocations", gain a 4KB region of C structs/arrays without a single pointer, linux ps/top mem count will drop a few kb.
